File: app/Shared/ProfileStore.swift
import Foundation

enum ProfileStore {
    private static var containerURL: URL? {
        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: HopConstants.appGroupID)
    }

    private static var storeURL: URL? {
        containerURL?.appendingPathComponent(HopConstants.profileStoreFileName)
    }

    static func load() -> AppState {
        guard let storeURL else {
            TunnelLog.error("App group unavailable — check entitlements for \(HopConstants.appGroupID)")
            return AppState()
        }
        guard FileManager.default.fileExists(atPath: storeURL.path) else {
            TunnelLog.info("No saved profile yet at \(storeURL.path)")
            return AppState()
        }
        guard let data = try? Data(contentsOf: storeURL) else {
            TunnelLog.error("Could not read profile at \(storeURL.path)")
            return AppState()
        }
        if var state = try? JSONDecoder().decode(AppState.self, from: data) {
            if state.chains.isEmpty, !state.servers.isEmpty {
                let id = state.addChain(name: "Default")
                if let index = state.chains.firstIndex(where: { $0.id == id }) {
                    state.chains[index].hopIDs = state.servers.map(\.id)
                }
                state.selectedChainID = id
                save(state)
            }
            return state
        }
        if let flat = try? JSONDecoder().decode(FlatHopsState.self, from: data) {
            TunnelLog.info("Migrated flat hop list to chains")
            return AppState.fromLegacyHops(flat.hops)
        }
        if let legacy = try? JSONDecoder().decode(LegacyHopLibrary.self, from: data) {
            TunnelLog.info("Migrated legacy hop library")
            return legacy.migrated()
        }
        TunnelLog.error("Profile decode failed at \(storeURL.path)")
        return AppState()
    }

    static func save(_ state: AppState) {
        guard let storeURL else {
            TunnelLog.error("App group save failed — entitlement \(HopConstants.appGroupID) missing?")
            return
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(state) else { return }
        do {
            try data.write(to: storeURL, options: .atomic)
        } catch {
            TunnelLog.error("Profile write failed: \(error.localizedDescription)")
        }
    }

    private static var lastErrorURL: URL? {
        containerURL?.appendingPathComponent(HopConstants.tunnelLastErrorFileName)
    }

    static func saveLastTunnelError(_ message: String) {
        guard let lastErrorURL else { return }
        try? message.write(to: lastErrorURL, atomically: true, encoding: .utf8)
    }

    static func loadLastTunnelError() -> String? {
        guard let lastErrorURL,
              let text = try? String(contentsOf: lastErrorURL, encoding: .utf8)
        else { return nil }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    static func clearLastTunnelError() {
        guard let lastErrorURL else { return }
        try? FileManager.default.removeItem(at: lastErrorURL)
    }
}

// MARK: - Legacy format migration

private struct FlatHopsState: Codable {
    var hops: [HopNodeProfile]
    var selectedHopID: UUID?
}

private struct LegacyHopLibrary: Codable {
    var nodes: [HopNodeProfile]
    var chains: [LegacyChain]
    var selectedChainID: UUID?

    struct LegacyChain: Codable {
        var id: UUID?
        var name: String?
        var hopIDs: [UUID]
    }

    func migrated() -> AppState {
        var state = AppState(servers: nodes)
        if chains.isEmpty, !nodes.isEmpty {
            return AppState.fromLegacyHops(nodes)
        }
        state.chains = chains.map {
            HopChain(
                id: $0.id ?? UUID(),
                name: $0.name ?? "",
                hopIDs: $0.hopIDs
            )
        }
        state.selectedChainID = selectedChainID ?? state.chains.first?.id
        return state
    }
}

File: app-android/app/src/main/java/com/aengix/hopper/data/ProfileStore.kt
package com.aengix.hopper.data

import android.content.Context
import com.aengix.hopper.model.AppState
import com.aengix.hopper.model.HopChain
import com.aengix.hopper.model.HopConstants
import com.aengix.hopper.model.HopNodeProfile
import com.aengix.hopper.util.TunnelLog
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

object ProfileStore {
    private lateinit var appContext: Context

    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = true
        encodeDefaults = true
    }

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    private val storeFile: File
        get() = File(appContext.filesDir, HopConstants.PROFILE_STORE_FILE_NAME)

    private val lastErrorFile: File
        get() = File(appContext.filesDir, HopConstants.TUNNEL_LAST_ERROR_FILE_NAME)

    fun load(): AppState {
        val file = storeFile
        if (!file.exists()) {
            TunnelLog.info("No saved profile yet at ${file.path}")
            return AppState()
        }

        val text = runCatching { file.readText() }.getOrElse {
            TunnelLog.error("Could not read profile at ${file.path}")
            return AppState()
        }

        runCatching { json.decodeFromString<AppState>(text) }.onSuccess { state ->
            if (state.chains.isEmpty() && state.servers.isNotEmpty()) {
                val migrated = AppState.fromLegacyHops(state.servers)
                save(migrated)
                return migrated
            }
            return state
        }

        runCatching { json.decodeFromString<FlatHopsState>(text) }.onSuccess { flat ->
            TunnelLog.info("Migrated flat hop list to chains")
            return AppState.fromLegacyHops(flat.hops)
        }

        runCatching { json.decodeFromString<LegacyHopLibrary>(text) }.onSuccess { legacy ->
            TunnelLog.info("Migrated legacy hop library")
            return legacy.migrated()
        }

        TunnelLog.error("Profile decode failed at ${file.path}")
        return AppState()
    }

    fun save(state: AppState) {
        runCatching {
            storeFile.writeText(json.encodeToString(state))
        }.onFailure {
            TunnelLog.error("Profile write failed: ${it.message}")
        }
    }

    fun saveLastTunnelError(message: String) {
        runCatching { lastErrorFile.writeText(message) }
    }

    fun loadLastTunnelError(): String? {
        if (!lastErrorFile.exists()) return null
        return lastErrorFile.readText().trim().ifEmpty { null }
    }

    fun clearLastTunnelError() {
        runCatching { lastErrorFile.delete() }
    }

    fun deviceId(): String {
        val file = File(appContext.filesDir, HopConstants.DEVICE_ID_KEY)
        if (file.exists()) {
            file.readText().trim().takeIf { it.isNotEmpty() }?.let { return it }
        }
        val fresh = java.util.UUID.randomUUID().toString()
        file.writeText(fresh)
        return fresh
    }
}

@Serializable
private data class FlatHopsState(
    val hops: List<HopNodeProfile>,
    val selectedHopID: String? = null,
)

@Serializable
private data class LegacyHopLibrary(
    val nodes: List<HopNodeProfile>,
    val chains: List<LegacyChain> = emptyList(),
    val selectedChainID: String? = null,
) {
    @Serializable
    data class LegacyChain(
        val id: String? = null,
        val name: String? = null,
        val hopIDs: List<String> = emptyList(),
    )

    fun migrated(): AppState {
        if (chains.isEmpty() && nodes.isNotEmpty()) {
            return AppState.fromLegacyHops(nodes)
        }
        val mappedChains = chains.map { chain ->
            HopChain(
                id = chain.id ?: java.util.UUID.randomUUID().toString(),
                name = chain.name.orEmpty(),
                hopIDs = chain.hopIDs,
            )
        }
        return AppState(
            servers = nodes,
            chains = mappedChains,
            selectedChainID = selectedChainID ?: mappedChains.firstOrNull()?.id,
        )
    }
}
